### ISSUE TITLE
api version is not optional

### ISSUE THREAD

**Issue by ahmettahasakar** [RSuter/NSwag#1891]
Hello,

I just upgraded to Nswag 12. The generated code as following:

before:
` index(): Observable<FileResponse | null> {`

after:
` index(api_version: string | null | undefined): Observable<FileResponse | null> {`

I get the following error:
error TS2554: Expected 1 arguments, but got 0.

Is this normal? Is there a way to inject api-version globally?

_Originally posted by @ahmettahasakar in https://github.com/RSuter/NSwag/issues/1601#issuecomment-444503860_

**Comment by ahmettahasakar** [comment]
I have found a workaround for now:

index(api_version? : string | null | undefined): Observable<FileResponse | null> {

When i add ?, it becomes optional.

I'm on typescript 3, by the way.

**Comment by RSuter** [comment]
Yes, it should generate a `?` in this case

**Comment by RSuter** [comment]
Have you tried to set `GenerateOptionalParameters` to `true`?

**Comment by ahmettahasakar** [comment]
😞 I got it. this is my swagger

```
 {
            "type": "string",
            "name": "api-version",
            "in": "header",
            "required": true,
            "x-nullable": true
          }
```

It shows required, because it's required by the api, but I'm setting it centrally with a http interceptor.

Is there a way to add a default api-version in the client with an InjectionToken?

**Comment by ahmettahasakar** [comment]
If there is no other solution, I will probably make it optional, but I rather have it set with an InjectionToken. What do you think @RSuter

**Comment by RSuter** [comment]
I think you should add a SwaggerIgnore attribute on the parameter so that it does not show up in the spec at all and then manually add the header in TS

**Comment by ahmettahasakar** [comment]
> I think you should add a SwaggerIgnore attribute on the parameter so that it does not show up in the spec at all and then manually add the header in TS

But it's a header that is controller by api version library. So I don't have a control over it

> Or we should add a ExcludedParameters setting in the code generator so optionally exclude some parameters from being generated (the global ones) so that they can be added manually...

I think this is best solution, and it will have other applications as well

**Comment by RSuter** [comment]
Maybe its not correctly setup, checkout: https://github.com/RSuter/NSwag/pull/1701

And https://github.com/RSuter/NSwag/issues/1355

**Comment by RSuter** [comment]
Can you create a new issue for ExcludedParameters?

**Comment by ahmettahasakar** [comment]
> Maybe its not correctly setup, checkout: #1701
> 
> And #1355

Yeah, `AssumeDefaultVersionWhenUnspecified` is false for me. I don't want it to be optional, but I want to be globally defined if not overridden. That's why I figured InjectionToken was a better choice, rather than excluding it completely. Perhaps I'm pushing too much and make it go beyond the standard approach

**Comment by RSuter** [comment]
As always, the goal is to support as many scenarios as possible with the least amount of settings... i think adding ExcludedParameters would be a nice addition.

You could also implement a global operation processor which removes all these params at spec generation time...
